### PR TITLE
Dove: Manifest for prover

### PR DESCRIPTION
**Pull request by vladimirovmm**

### CODE REVIEW

File: dove/src/manifest.rs [hunk 1]
```diff
@@ -34,10 +36,55 @@ pub struct DoveToml {
     /// Documentation generator operations.
     #[serde(default)]
     pub doc: DocgenOptions,
+    /// Boogie Options
+    pub boogie_options: Option<BoogieOptions>,
+}
+
+impl PartialEq for DoveToml {
```
Comment (RIg410) [inline dove/src/manifest.rs:43]: What is it for ?

File: dove/src/cmd/prover.rs [hunk 2]
```diff
@@ -83,18 +138,57 @@ fn is_z3_available(z3_exe: &str) -> bool {
...
+
+/// get path to z3
+fn get_path_z3() -> Result<String, anyhow::Error> {
+    let env_path = env!("PATH");
+
+    #[cfg(not(target_family = "windows"))]
+    let separator = ':';
+
+    #[cfg(target_family = "windows")]
+    let separator = ':';
```
Comment (vitvakatu) [inline dove/src/cmd/prover.rs:172]: Same separator for both feature flags

File: dove/src/cmd/prover.rs [hunk 3]
```diff
@@ -83,18 +138,57 @@ fn is_z3_available(z3_exe: &str) -> bool {
...
+                println!("Warning: {}", String::from_utf8(result.stderr).unwrap());
+            }
+            result.status.success()
+        }
         Err(_) => false,
     }
 }
+
+/// get path to z3
+fn get_path_z3() -> Result<String, anyhow::Error> {
```
Comment (vitvakatu) [inline dove/src/cmd/prover.rs:165]: Why `z3` is the only one that gets a special method while others just use simple constants?

File: dove/src/manifest.rs (hunk 1 above)
Comment (vitvakatu) [reply dove/src/manifest.rs:43]: Yeah, we shouldn't do it that way

File: dove/src/cmd/prover.rs [hunk 4]
```diff
@@ -83,18 +138,57 @@ fn is_z3_available(z3_exe: &str) -> bool {
...
+    let separator = ':';
+
+    env_path
+        .split(separator)
+        .map(PathBuf::from)
+        .map(|path| path.join(Z3_EXE))
+        .find(|path|path.exists())
+        .map_or(
+            Err(anyhow!("z3 executable not found in PATH. Please install it from https://github.com/Z3Prover/z3")),
+                |path| path.to_str().ok_or_else(||anyhow!("couldn't convert path to string")).map(|t|t.to_string())
```
Comment (vitvakatu) [inline dove/src/cmd/prover.rs:181]: Rustfmt is too forgiving on this line...

File: dove/src/manifest.rs [hunk 5]
```diff
@@ -171,27 +222,38 @@ pub struct Layout {
...
     #[serde(default = "storage_dir")]
     pub storage_dir: String,
+
     /// Path to executor build directory.
     #[serde(default = "exe_build_dir")]
     pub exe_build_dir: String,
+
+    /// Path to pover settings
+    #[serde(default = "prover_toml")]
+    pub prover_toml: String,
```
[the remark this answers is not in the document]
Comment (vladimirovmm) [reply dove/src/manifest.rs:256]: Thank you for your comment. This was the task

File: dove/src/cmd/prover.rs (hunk 3 above)
Comment (vladimirovmm) [reply dove/src/cmd/prover.rs:165]: @vitvakatu He's the only one who needs a path

File: dove/src/manifest.rs (hunk 1 above)
Comment (vladimirovmm) [reply dove/src/manifest.rs]: Thanks. Fixed

File: dove/src/cmd/prover.rs [hunk 6]
```diff
@@ -74,6 +89,46 @@ impl Cmd for Prove {
...
     }
 }
+impl Prove {
+    fn get_prover_conf(&self, ctx: &Context) -> Result<ProverConf, anyhow::Error> {
+        // cmd
+        let mut boogie_exe = self.boogie_exe.clone();
+        let mut z3_exe = self.z3_exe.clone();
+        let mut cvc4_exe = self.cvc4_exe.clone();
+        // prover-env.toml
+        if boogie_exe.is_none() || z3_exe.is_none() || cvc4_exe.is_none() {
```
[the remark this answers is not in the document]
Comment (vladimirovmm) [reply dove/src/cmd/prover.rs]: @vitvakatu Thanks. Fixed

File: dove/src/cmd/prover.rs [hunk 7]
```diff
@@ -33,17 +37,27 @@ pub struct Prove {
...
-        let z3_exe = self.z3_exe.unwrap_or_else(|| Z3_EXE.to_string());
+        let (boogie_exe, z3_exe, cvc4_exe) = self.get_prover_conf(&ctx).map(|conf| {
+            (
+                conf.boogie_exe.unwrap_or_default(),
+                conf.z3_exe.unwrap_or_default(),
+                conf.cvc4_exe.unwrap_or_default(),
+            )
+        })?;
+
+        ensure!(is_boogie_available(&boogie_exe), "boogie executable not found in PA    TH. Please install it from https://github.com/boogie-org/boogie");
```
[the remark this answers is not in the document]
Comment (vladimirovmm) [reply dove/src/cmd/prover.rs]: @vitvakatu Thanks. Fixed

File: dove/src/cmd/prover.rs [hunk 8]
```diff
@@ -74,6 +89,46 @@ impl Cmd for Prove {
         run_move_prover_errors_to_stderr(options, &mut preprocessor)
     }
 }
+impl Prove {
+    fn get_prover_conf(&self, ctx: &Context) -> Result<ProverConf, anyhow::Error> {
```
[the remark this answers is not in the document]
Comment (vladimirovmm) [reply dove/src/cmd/prover.rs]: @vitvakatu Thanks. Fixed

File: dove/src/cmd/prover.rs [hunk 9]
```diff
@@ -83,18 +138,57 @@ fn is_z3_available(z3_exe: &str) -> bool {
...
+        .find(|path|path.exists())
+        .map_or(
+            Err(anyhow!("z3 executable not found in PATH. Please install it from https://github.com/Z3Prover/z3")),
+                |path| path.to_str().ok_or_else(||anyhow!("couldn't convert path to string")).map(|t|t.to_string())
+            )
+}
+
+/// Paths to the boogie, z3 and cvc4 binaries
+#[derive(Debug, Deserialize)]
+struct ProverConf {
```
[the remark this answers is not in the document]
Comment (vladimirovmm) [reply dove/src/cmd/prover.rs]: @vitvakatu Thanks. Fixed

File: dove/src/cmd/prover.rs (hunk 2 above)
Comment (vladimirovmm) [reply dove/src/cmd/prover.rs]: @vitvakatu Thanks. Fixed

File: dove/src/cmd/prover.rs (hunk 4 above)
Comment (vladimirovmm) [reply dove/src/cmd/prover.rs]: @vitvakatu Thanks. Fixed

**Review by vitvakatu — APPROVED** [review]

**Review by RIg410 — APPROVED** [review]

**Pull request merged**